README.md
# Winmin Tools

Winmin is a set of scripts and tools for installing, managing, and running Windows on Linux applications using [libvirt](https://libvirt.org/) virtual machines.

# Build Dependencies

- python3

# Building

Build and install with:

```
$ meson build
$ ninja -C build install
```
# Runtime Dependencies

- python3
- python3-magic
- python-is-python3
- libvirt-clients
- libvirt-daemon-system
- virtinst
- samba
- [winmin-viewer](https://github.com/vlinkz/winmin-viewer)

# Setup

In order to use Winmin, you need to set up a base VM using [winmin-setup](https://github.com/vlinkz/winmin-setup).
To access the spice socket of the guest VM, the user must be part of the `kvm` group. This can be added using the following command.
```
$ sudo usermod -aG $USER kvm
```
In order to interact with the guest serial port, the user must be part of the `tty` group. This can be added using the following command.
```
$ sudo usermod -aG $USER tty
```
You may need to logout or reboot in order for group changes to take effect.

In order to transfer files between the host and guest, a samba server must be set up. To set this up, add the following to the end of `/etc/samba/smb.conf`.
```
[winmin]
  comment = VM Filesystem
  path = /
  browseable = yes
  read only = no
```

After changing the samba config, add an smb account for your user.
```
$ sudo smbpasswd -a $USER
```

Then restart the samba service.

```
$ sudo systemctl restart smbd
```
Enable the service if not automatically enabled
```
$ sudo systemctl enable smbd
```

# Running

## winmin-install

Installs a package given an exe or msi file. This will create a new vm with the base created using [winmin-setup](https://github.com/vlinkz/winmin-setup) as a backing file. The `--web` flag will open a link in the guest's default web browser where a user can manually download an installer (This is useful for applications that require sign-in to download).

### Usage
```
winmin-install [-h] [--web] input shortname
```
### Examples
```
winmin-install ./vs_community.exe visualstudiocommunity
winmin-install --web https://office.com/ ms-office
```
## winmin-yml-install

This tool installs a package given a Winmin `.yaml` package file. This tool also automatically creates `.desktop` files for the packages specefied within the `.yaml` files.

### Usage
```
winmin-yml-install [-h] input
```
### Examples
```
winmin-yml-install ./ms-office.yaml
```

## winmin-run

This tool runs an installed application in the specified Winmin VM. By default, the VM is destroyed once its window is closed. Using the `--save` flag changes this so the VM is saved instead (Saving takes a longer time, and attempting to open an application while its VM is being saved will not work). This tool can be run with input file.

### Usage
```
winmin-run [-h] [--save] vm program name [input]
```

### Examples
```
winmin-run winmin-msoffice Word Word ./File.docx
winmin-run winmin-msoffice Outlook Outlook --save
```

# TODO
- Add direct binding to libvirt virtual machines using the python libvirt API
- Add a yaml file generator similar to that of [winget](https://github.com/microsoft/winget-pkgs/blob/master/Tools/YamlCreate.ps1)
- Fix install file logic in [winmin-yml-install](https://github.com/vlinkz/winmin-tools/blob/b44cc6bfcd7625db4d5b5004ac26bc0e1a5c1572/winmin-scripts/winmin_yml_install.py#L49)
- Add options for more descriptive `.desktop` files in `.yaml` files
- Add and test arument support (other than input files)

debian/changelog
winmin-tools (0.1-1) unstable; urgency=low

  * Initial release

 -- Victor Fuentes <hyruleterminatriforce@gmail.com>  Mon, 02 Jun 2020 00:20:56 +0100



winmin.c
#include <gtk/gtk.h>
#include <spice-client-gtk-3.0/spice-client-gtk.h>
#include <stdio.h>
SpiceMainChannel * mainchannel;
GtkWidget *window;
char** name;
char** sock;
int id;

gboolean resize_done (gpointer data)
{
  guint *gid = data;
  *gid = 0;
  resize(window);
  return FALSE;
}


gboolean on_configure_event (GtkWidget *window, GdkEvent *event, gpointer data)
{
  static guint gid = 0;
  if (gid)
    g_source_remove (gid);
  gid = g_timeout_add (250, resize_done, &gid);
  return FALSE;
}

void resize(GtkWidget * window) {
  printf("Resize!\n");
  int width, height;
  gtk_window_get_size(window, &width, &height);
  spice_main_channel_update_display(mainchannel, id, 0, 0, width, height, TRUE);
  spice_main_update_display_enabled(mainchannel,id,TRUE,TRUE);
  spice_main_send_monitor_config(mainchannel);
}

static void
activate (GtkApplication* app,
          gpointer        user_data)
{
  
  window = gtk_application_window_new (app);

  SpiceSession * session = spice_session_new();
  g_object_set(session,"uri",sock);
     //"spice+unix:///var/lib/libvirt/qemu/domain-1-winmin-base/spice.sock");
     //0.0.0.0:5900");

  SpiceChannel * channel =  spice_channel_new(session,1,id); // SPICE_CHANNEL(session);
  if (SPICE_IS_CHANNEL(channel)) {
    printf("Is spice channel\n");
  }
  else {
    printf("Not spice channel\n");
    exit(1);
  }

  printf("ID: %d\n",id);

  spice_session_connect(session);
  mainchannel =  SPICE_MAIN_CHANNEL(channel);
  if (SPICE_IS_MAIN_CHANNEL(mainchannel)) {
    printf("Good\n");
  }
  else {
    printf("Bad\n");
  }

  SpiceDisplay * display = spice_display_new(session,0);
  

  GtkWidget * displaywidget = GTK_WIDGET(display);
  gtk_container_add(GTK_CONTAINER(window),displaywidget);
    
  gtk_window_set_title (GTK_WINDOW(window), name);
  gtk_window_set_default_size (GTK_WINDOW (window), 900, 800);
  gtk_window_set_position(GTK_WINDOW(window), GTK_WIN_POS_CENTER);
  gtk_widget_show_all (window);

  resize(window);
  g_signal_connect(G_OBJECT(window), "configure-event",
        G_CALLBACK(on_configure_event), NULL);
}

int
main (int    argc,
      char **argv)
{
  GtkApplication *app;
  int status;
  name = argv[1];
  sock = argv[2];
  id = 0; //(int)strtol(argv[3], NULL, 10);
  app = gtk_application_new(NULL, G_APPLICATION_FLAGS_NONE); //"org.vlinkz.winmin"
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  status = g_application_run (G_APPLICATION (app), NULL, NULL);
  g_object_unref (app);

  return status;
}


winmin-run.py
#!/usr/bin/env python3
import os,subprocess,sys
import time
import argparse
import socket

def startup(vm,program,arg,name,save):

  print("Starting")
  subprocess.call("virsh restore /var/lib/libvirt/qemu/save/{}.save --running".format(vm),shell=True)

  dump = subprocess.check_output("virsh dumpxml {}".format(vm),shell=True).decode("utf-8").split("'")
  print("Finding SPICE socket")
  for i in range(len(dump)):
    if "spice.sock" in dump[i]:
      sock="spice+unix://{}".format(dump[i])
      print(sock)
      break

  print("Checking pts")
  pts = subprocess.check_output("virsh ttyconsole {}".format(vm),shell=True).decode('UTF-8').strip('\n\n')
  print("Starting serial")
  with open(pts, 'w') as p:
    p.write("\r\r")
    if len(arg) >= 1:
      p.write('startps "{}" "{}"\r'.format(program,arg))
    else:
      p.write('startps "{}"\r'.format(program))
    time.sleep(1)
    p.write("\r\r")
  print("commands EXECUTED!!!")
  #time.sleep(2)
  subprocess.call('/home/victor/bin/winmin2 "{}" "{}"'.format(name,sock),shell=True)

  end(vm)

def end(vm):
  subprocess.call("virsh send-key {} KEY_LEFTALT KEY_F".format(vm),shell=True)
  subprocess.call("virsh send-key {} KEY_X".format(vm),shell=True)
  time.sleep(0.5)
  subprocess.call("virsh send-key {} KEY_LEFTALT KEY_F4".format(vm),shell=True)
  time.sleep(1)

  if save:
    subprocess.call("virsh managedsave {}".format(vm),shell=True)
  else:
    subprocess.call("virsh destroy {}".format(vm),shell=True)



parser = argparse.ArgumentParser(description='Run an application to WinMin.')

parser.add_argument("vm", type=str, help="WinMin vm to run")
parser.add_argument("program", type=str, help="Program to run in WinMin")
parser.add_argument("name", type=str, help="Name of program to run")
parser.add_argument("input", type=str, nargs="?",help="Input a file for a program to open")
parser.add_argument("--save", action="store_true", help="Save the vm after closing")

args=parser.parse_args()

program=args.program
save=args.save
name=args.name
vm=args.vm
if type(args.input) != type(None) :
  file=os.path.realpath(args.input)
  s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
  s.connect(("8.8.8.8", 0))
  ip = s.getsockname()[0]
  file="\\\\{}\\winmin".format(ip)+file.replace("/","\\")
else:
  file=""


startup(vm,program,file,name,save)


winmin-install.py
#!/usr/bin/env python3
import os,sys,subprocess
import argparse
import time
import magic

def createvm(vm):

  subprocess.call("sudo qemu-img create -f qcow2 -F raw -b /var/lib/libvirt/images/win10base.img /var/lib/libvirt/images/{}.qcow2".format(vm),shell=True)
  subprocess.call("virt-install --virt-type=kvm --name={} --ram 4096 --vcpus 4 --hvm --network network=default,model=virtio --graphics spice,listen=socket --disk /var/lib/libvirt/images/{}.qcow2,bus=virtio --boot=hd --video qxl --noautoconsole --os-variant win10".format(vm,vm),shell=True)

  print("Booting new vm")
  time.sleep(32)

  pts = subprocess.check_output("virsh ttyconsole {}".format(vm),shell=True).decode('UTF-8').strip('\n\n')
  time.sleep(0.5)
  print("Setting up serial interface")

  with open(pts, 'w') as p:
    time.sleep(0.5)
    p.write("\n")
    time.sleep(0.5)
    p.write("cmd\n")
    time.sleep(1.5)
    p.write("ch -sn Cmd0001\n")
    time.sleep(0.5)
    p.write("\r")
    time.sleep(0.5)
    p.write("vm\r")
    time.sleep(0.5)
    p.write("\r")
    time.sleep(0.5)
    p.write("vm\r")
    time.sleep(0.5)
    p.write("\r")

  return pts

def installapp(file,vm,pts):
  print("Starting installer")
  filename = file.split("\\")[-1]
  print(filename)

  with open(pts, 'w') as p:
    p.write("\r")
    time.sleep(0.5) #wait for shell to start
    p.write("dir C:\\ \r") #Idk why but it works after buffering a command
    time.sleep(1)
    p.write("    copy {} C:\\Users\\VM\\Downloads\\ /Y \r".format(file))
    time.sleep(0.5)
    p.write("startps C:\\Users\\VM\\Downloads\\{} \r".format(filename))
    time.sleep(0.5)
    p.write("\r\r")
  time.sleep(2)

  dump = subprocess.check_output("virsh dumpxml {}".format(vm),shell=True).decode("utf-8").split("'")
  print("Finding SPICE socket")
  for i in range(len(dump)):
    if "spice.sock" in dump[i]:
      sock="spice+unix://{}".format(dump[i])
      print(sock)
      break

  subprocess.call("/home/victor/bin/winmin2 \"WinMin Installer\" \"{}\"".format(sock),shell=True)
  subprocess.call("virsh managedsave {}".format(vm),shell=True)

'''
def setupdesktop(name,program):
  template = open("/home/victor/bin/template.desktop","rt")
  out = open("/home/victor/.local/share/applications/visual-studio-2019.desktop", 'wt')
  for line in template:
    out.write(line.replace("{{NAME}}",name))
    out.write(line.replace("{{PROGRAM}}",program))
  template.close()
  out.close()
'''

parser = argparse.ArgumentParser(description='Install application to WinMin.')
parser.add_argument('input', help='input file path')
parser.add_argument('shortname', help='Name of vm being created')
#parser.add_argument('programs', nargs="?", help='Programs being installed (needs to be exact name windows uses in the start menu)')
args = parser.parse_args()

filetypes=["application/x-msi","application/x-dosexec"]

if magic.from_file(args.input,mime=True) not in filetypes:
  print("Input is not a valid Windows application installer, exiting")
  exit()

file=os.path.realpath(args.input)
file="\\\\192.168.2.30\\winmin"+file.replace("/","\\")
shortname=args.shortname
#shortname=str.lower(name).strip(" ")
#if len(args.program) > 0:
#program=args.program
#else:
#  program=name
vm="winmin-{}".format(shortname)


pts = createvm(vm)
#pts = subprocess.check_output("virsh ttyconsole winmin-VS2k19",shell=True).decode('UTF-8').strip('\n\n')
time.sleep(1)
installapp(file,vm,pts)
#setupdesktop(name,program)


winmin-yml-install.py
import yaml
import os,sys,subprocess
import argparse
import urllib.request


def installdesk(data):
  programs=data["programs"]
  for i in range(len(programs)):
    program = programs["program{}".format(i+1)]
    name = program["name"]
    shortname = "winmin-"+str.lower(name).replace(" ","-")
    iconurl = program["icon"]["location"]
    iconname = "winmin-"+str.lower(iconurl.split("/")[-1]).replace(" ","-")
    urllib.request.urlretrieve(iconurl, '/home/victor/.local/share/pixmaps/{}'.format(iconname))
    print("shortname: {}".format(shortname))

    template = open("/home/victor/bin/template.desktop","rt")
    out = open("/home/victor/.local/share/applications/{}.desktop".format(shortname), 'wt')
    for line in template:
      out.write(line.replace("{{NAME}}",name).replace("{{PROGRAM}}",name).replace("{{VM}}","winmin-"+data["shortname"]).replace("{{ICON}}","/home/victor/.local/share/pixmaps/{}".format(iconname)))




parser = argparse.ArgumentParser(description='Install application to WinMin using a yaml file.')
parser.add_argument('input', help='input yaml file path')
args = parser.parse_args()
file=args.input

if not os.path.isfile(file):
  print("File not found")
  exit()
if os.path.exists("/tmp/winmin-tmp"):
  os.system("rm -r /tmp/winmin-tmp")

with open(file) as f:
  data = yaml.load(f, Loader=yaml.FullLoader)
  print(data)

os.mkdir("/tmp/winmin-tmp")

shortname = data["shortname"]
url = data["url"]
filename = url.split("/")[-1]

urllib.request.urlretrieve(url, '/tmp/winmin-tmp/{}'.format(filename))
file="/tmp/winmin-tmp/{}".format(filename)

subprocess.call("/home/victor/bin/winmin-install \"{}\" \"{}\"".format(file,shortname),shell=True)

installdesk(data)

#os.system("rm -r /tmp/winmin-tmp")

winmin-scripts/winmin_run.py
#!/usr/bin/python3
import os,subprocess,sys
import time
import argparse
import socket

def startup(vm,program,arg,name,save):

  print("Starting")
  subprocess.call("virsh restore /var/lib/libvirt/qemu/save/{}.save --running".format(vm),shell=True)

  dump = subprocess.check_output("virsh dumpxml {}".format(vm),shell=True).decode("utf-8").split("'")
  print("Finding SPICE socket")
  for i in range(len(dump)):
    if "spice.sock" in dump[i]:
      sock="spice+unix://{}".format(dump[i])
      print(sock)
      break

  print("Checking pts")
  pts = subprocess.check_output("virsh ttyconsole {}".format(vm),shell=True).decode('UTF-8').strip('\n\n')
  print("Starting serial")
  with open(pts, 'w') as p:
    p.write("vm\r") #In case logged out in previous session
    p.write("\r")
    time.sleep(0.5)
    if len(arg) >= 1:
      p.write('startps "{}" "{}"\r'.format(program,arg))
    else:
      p.write('startps "{}"\r'.format(program))
    time.sleep(1)
    p.write("\r\r")
  print("commands EXECUTED!!!")
  #time.sleep(2)
  subprocess.call('winmin-viewer "{}" "{}"'.format(name,sock),shell=True)

  end(vm,save)

def end(vm,save):
  subprocess.call("virsh send-key {} KEY_LEFTALT KEY_F".format(vm),shell=True)
  subprocess.call("virsh send-key {} KEY_X".format(vm),shell=True)
  time.sleep(0.5)
  subprocess.call("virsh send-key {} KEY_LEFTALT KEY_F4".format(vm),shell=True)
  time.sleep(1)

  if save:
    subprocess.call("virsh managedsave {}".format(vm),shell=True)
  else:
    subprocess.call("virsh destroy {}".format(vm),shell=True)


def main():
  parser = argparse.ArgumentParser(description='Run an application to WinMin.')

  parser.add_argument("vm", type=str, help="WinMin vm to run")
  parser.add_argument("program", type=str, help="Program to run in WinMin")
  parser.add_argument("name", type=str, help="Name of program to run")
  parser.add_argument("input", type=str, nargs="?",help="Input a file for a program to open")
  parser.add_argument("--save", action="store_true", help="Save the vm after closing")

  args=parser.parse_args()

  program=args.program
  save=args.save
  name=args.name
  vm=args.vm
  if type(args.input) != type(None) :
    file=os.path.realpath(args.input)
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect(("8.8.8.8", 0))
    ip = s.getsockname()[0]
    file="\\\\{}\\winmin".format(ip)+file.replace("/","\\")
  else:
    file=""
  
  startup(vm,program,file,name,save)
  
if __name__ == "__main__":
  main();


winmin-scripts/winmin_yml_install.py
#!/usr/bin/python3
import yaml
import os,sys,subprocess
import argparse
import urllib.request


def installdesk(data):
  home=os.environ['HOME']
  for program in data["Programs"]:
    name = program["Name"]
    shortname = "winmin-"+str.lower(name).replace(" ","-")
    if program["Icon"]["Link"]:
      iconurl = program["Icon"]["Location"]
      iconname = "winmin-"+str.lower(iconurl.split("/")[-1]).replace(" ","-")
      if not os.path.exists("{}/.local/share/pixmaps/".format(home)):
        os.mkdir("{}/.local/share/pixmaps/".format(home))
      urllib.request.urlretrieve(iconurl, '{}/.local/share/pixmaps/{}'.format(home,iconname))
      icon = "{}/.local/share/pixmaps/{}".format(home,iconname)
    else:
      icon = program["Icon"]["Location"]
    print("shortname: {}".format(shortname))

    template = open("/usr/share/winmin/template.desktop","rt")
    out = open("{}/.local/share/applications/{}.desktop".format(home,shortname), 'wt')
    for line in template:
      out.write(line.replace("{{NAME}}",name).replace("{{PROGRAM}}",name).replace("{{VM}}","winmin-"+data["Id"]).replace("{{ICON}}",icon))

def main():
  parser = argparse.ArgumentParser(description='Install application to WinMin using a yaml file.')
  parser.add_argument('input', help='input yaml file path')
  args = parser.parse_args()
  file=args.input

  if not os.path.isfile(file):
    print("File not found")
    exit()
  if os.path.exists("/tmp/winmin-tmp"):
    os.system("rm -r /tmp/winmin-tmp")

  with open(file) as f:
    data = yaml.load(f, Loader=yaml.FullLoader)
    print(data)

  os.mkdir("/tmp/winmin-tmp")
  
  shortname = data["Id"]
  url = data["Installer"]["Url"]
  filename = url.split("/")[-1]

  if data["Installer"]["InstallerType"] == "web":
    subprocess.call("winmin-install \"{}\" \"{}\" --web".format(url,shortname),shell=True)
  else: #TODO, specify if executing exe, and if user input is necessary (see winget repo for ideas)
    urllib.request.urlretrieve(url, '/tmp/winmin-tmp/{}'.format(filename))
    installfile="/tmp/winmin-tmp/{}".format(filename)
    subprocess.call("winmin-install \"{}\" \"{}\"".format(installfile,shortname),shell=True)
  
  installdesk(data)
    
  #os.system("rm -r /tmp/winmin-tmp")
 
if __name__ == "__main__":
  main();
 


winmin-scripts/winmin_install.py
#!/usr/bin/python3
import os,sys,subprocess
import argparse
import time
import magic
import socket

def createvm(vm):

  subprocess.call("sudo qemu-img create -f qcow2 -F raw -b /var/lib/libvirt/images/winmin-base.img /var/lib/libvirt/images/{}.qcow2".format(vm),shell=True)
  subprocess.call("virt-install --virt-type=kvm --name={} --ram 4096 --vcpus 4 --hvm --network network=default,model=virtio --graphics spice,listen=socket --disk /var/lib/libvirt/images/{}.qcow2,bus=virtio,cache=none,io=native --boot=hd --video qxl --noautoconsole --os-variant win10".format(vm,vm),shell=True)

  print("Booting new vm")
  time.sleep(20)

  pts = subprocess.check_output("virsh ttyconsole {}".format(vm),shell=True).decode('UTF-8').strip('\n\n')
  time.sleep(0.5)
  print("Setting up serial interface")

  with open(pts, 'w') as p:
    time.sleep(0.5)
    p.write("\n")
    time.sleep(0.5)
    p.write("cmd\n")
    time.sleep(1.5)
    p.write("ch -sn Cmd0001\n")
    time.sleep(0.5)
    p.write("\r")
    time.sleep(0.5)
    p.write("vm\r")
    time.sleep(0.5)
    p.write("\r")
    time.sleep(0.5)
    p.write("vm\r")
    time.sleep(0.5)
    p.write("\r")

  return pts

def installapp(file,vm,pts):
  print("Starting installer")
  filename = file.split("\\")[-1]
  print(filename)

  with open(pts, 'w') as p:
    p.write("\r")
    time.sleep(0.5) #wait for shell to start
    p.write("dir C:\\ \r") #Idk why but it works after buffering a command
    time.sleep(1)
    p.write("    copy {} C:\\Users\\VM\\Downloads\\ /Y \r".format(file))
    time.sleep(0.5)
    p.write("startps C:\\Users\\VM\\Downloads\\{} \r".format(filename))
    time.sleep(0.5)
    p.write("\r\n\r")
  time.sleep(2)

  dump = subprocess.check_output("virsh dumpxml {}".format(vm),shell=True).decode("utf-8").split("'")
  print("Finding SPICE socket")
  for i in range(len(dump)):
    if "spice.sock" in dump[i]:
      sock="spice+unix://{}".format(dump[i])
      print(sock)
      break

  subprocess.call("winmin-viewer \"WinMin Installer\" \"{}\"".format(sock),shell=True)
  subprocess.call("virsh managedsave {}".format(vm),shell=True)


def userinstall(site,vm,pts):
  print("Starting web installer")

  with open(pts, 'w') as p:
    p.write("\r")
    time.sleep(0.5) #wait for shell to start
    p.write("dir C:\\ \r") #Idk why but it works after buffering a command
    time.sleep(1)
    p.write("    startps microsoft-edge:{} \r".format(site))
    time.sleep(0.5)
    p.write("\r\n\r")
  time.sleep(2)

  dump = subprocess.check_output("virsh dumpxml {}".format(vm),shell=True).decode("utf-8").split("'")
  print("Finding SPICE socket")
  for i in range(len(dump)):
    if "spice.sock" in dump[i]:
      sock="spice+unix://{}".format(dump[i])
      print(sock)
      break

  subprocess.call("winmin-viewer \"WinMin Installer\" \"{}\"".format(sock),shell=True)
  subprocess.call("virsh managedsave {}".format(vm),shell=True)

def main():
  parser = argparse.ArgumentParser(description='Install application to WinMin.')
  parser.add_argument('input', help='input file path or url')
  parser.add_argument('shortname', help='Name of vm being created')
  parser.add_argument("--web", action="store_true", help="Open install link in web browser")

  #parser.add_argument('programs', nargs="?", help='Programs being installed (needs to be exact name windows uses in the start menu)')
  args = parser.parse_args()

  if not args.web:
    filetypes=["application/x-msi","application/x-dosexec"]
    if magic.from_file(args.input,mime=True) not in filetypes:
      print("Input is not a valid Windows application installer, exiting")
      exit()
    file=os.path.realpath(args.input)
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect(("8.8.8.8", 0))
    ip = s.getsockname()[0]
    file="\\\\{}\\winmin".format(ip)+file.replace("/","\\")
  
  shortname=args.shortname
  vm="winmin-{}".format(shortname)


  pts = createvm(vm)
  time.sleep(1)
  if args.web:
    userinstall(args.input,vm,pts)
  else:
    installapp(file,vm,pts)

if __name__ == "__main__":
  main();
